### PR TITLE
Switch to Slack App

### PR DESCRIPTION
**Pull request by mgwalker**
This PR refactors Charlie to use the Bolt API created by Slack instead of Hubot. This allows us to make Charlie a Slack App instead of a legacy API integration.

In the process of migrating, I also wrapped up some common functions into utilities. There are Slack utilities (e.g., adding a reaction emoji or posting an ephemeral message), Tock utilities (i.e., fetching a list of truant 18F Tock users), and a cache utility. I also switched from Mocha to Jest testing framework, since it seems to have a more modern lean, and allowed me to remove some other plugins and rely on the functionality that Jest has built-in.

Finally, I went through the documentation and updated it accordingly, trying to make sure to finally accurately capture the list of environment variables that Charlie understands. I also created a list of Slack OAuth scopes and events that Charlie needs in order to function.

### CODE REVIEW

**Comment by afeld** [comment]
Closes https://github.com/18F/18f-bot/issues/235. I put in a support request to disconnect from the upstream, regardless.

File: env.md
Comment (mgwalker) [inline env.md:1]: Split environment variables and Slack OAuth configuration into separate files. This is the first of those.

File: src/brain.js
Comment (mgwalker) [inline src/brain.js:4]: Previously, we used Hubot which had a `brain` plugin that stored stuff in redis. Since we needed to reimplement this "brain" behavior, I switched us to postgres on the assumption it would be cheaper to maintain.

**Review by mgwalker — COMMENTED** [review]
Not all of the files that the diff claims were deleted actually were; many were moved. The diff didn't pick that up, though, so there's more to look at than actually changed. Overall, I think this is ready to go, though.

**Comment by mgwalker** [comment]
@afeld This is ready for a code review and I'll set about finding someone to take care of that, but we can't merge it until the Slack app is created with the necessary credentials, scopes, and events configured. The necessary Slack configuration is documented in [the oath.md file](https://github.com/18F/charlie/blob/switch-to-bolt/oauth.md) in this PR.

Just a heads up for you that the code is ready to move on this whenever your team gets around to it. 😄

File: src/utils/tock.js
Comment (Sgtpluck) [inline src/utils/tock.js:35]: 😭 (i think i sobbed about this on another PR)

**Review by Sgtpluck — APPROVED** [review]
This is very cool! i have some suggestions and comments but none of them are blockers. 

lots of suggestions on the new tests but i am happy to go do a clean-up pr for them if you don't want to update them in this PR; it's very exciting to have them!

i tried to leave pseudocode when making suggestions but please ask if anything is unclear or if there are questions about suggestions themselves. thanks for doing this!

File: src/include.test.js
Comment (mgwalker) [reply src/include.test.js:6]: Honestly, I did this to make sure it pulled in the files that didn't have tests. That way I could glance at the coverage report to see which ones I still needed to do. 🙂 And now when new scripts get added, they'll show up in the coverage report immediately, even if they don't have tests yet.

File: src/scripts/angry-tock.test.js
Comment (mgwalker) [reply src/scripts/angry-tock.test.js:157]: This one is quirky because it's _only_ a holiday for them. Shouty-bot is not location-aware, but that could be a neat enhancement.

Also, I have to admit I only picked this date because I was looking for a significant Monday between 1991 and 2019 (so it fit chronologically between the previous and following tests). I picked 1997 at random and started scanning Wikipedia. My criteria was a Monday that was not a holiday where the next Monday was also not a holiday, and relatively significant events happened on both Mondays. I picked the first one I found. 😂

File: src/scripts/angry-tock.test.js
Comment (mgwalker) [reply src/scripts/angry-tock.test.js:181]: It's a Friday! Shouty-bot shouts on Mondays. 😄

File: src/scripts/angry-tock.test.js
Comment (mgwalker) [reply src/scripts/angry-tock.test.js:314]: I knoooowwwww! As I was porting this one over, I realized it was probably not good tests. But I didn't want to take the time to refactor them right then and figured I'd do it later. Like, it's testing too many things! It's not isolating very well.

File: src/scripts/coffeemate.test.js
Comment (mgwalker) [reply src/scripts/coffeemate.test.js:89]: Only slightly. It just verifies that the bot reads from the "brain" when it initializes. But we could just test for that exact thing, since the behavior afterwards is the same as above.

File: src/scripts/federal-holidays-reminder.js
Comment (mgwalker) [reply src/scripts/federal-holidays-reminder.js:19]: I am ashamed. There are libraries that will do truthy-type checks for strings and I should just use one of those. 😬  I will do that. OR, I may just take out the `SUPPRESS_HERE` altogether, since we don't actually use it at all. It was a premature optimization to ease some concerns that never actually materialized.

File: src/scripts/federal-holidays-reminder.js
Comment (mgwalker) [reply src/scripts/federal-holidays-reminder.js:67]: Ah, so this is a trick so I can mock `getNextHoliday` in the tests. Without this, you end up mocking the module export, which turns into a different pointer, and so you can't actually prevent `scheduleReminder` from calling the real `getNextHoliday`. It's kind of ugly, but it simplifies the tests.

That said... this is another one where I don't like the tests. They are testing too much internal behavior instead of just testing the in/out. 😕

File: src/scripts/federal-holidays-reminder.test.js
Comment (mgwalker) [reply src/scripts/federal-holidays-reminder.test.js:48]: This is one of those tests that, IMO, should go away because it's internal functionality. But yeah, these tests are hard to think about.

File: src/scripts/federal-holidays.js
Comment (mgwalker) [reply src/scripts/federal-holidays.js:14]: 🧠  YES! These used to be separate libraries so they couldn't share code, but now they're all right in the bot so they totally should!

File: src/scripts/love.js
Comment (mgwalker) [reply src/scripts/love.js:15]: Agreed, but `prettier` autoformats them this way. I think it's something to do with the depth of destructuring, but I'm not sure.

File: src/scripts/love.js
Comment (mgwalker) [reply src/scripts/love.js:40]: Only `channelID` could be undefined. `lovees` is guaranteed by the regex in the event subscription, `permalink` and `user` are guaranteed by the Slack API (because we got the inputs from Slack).

It won't fail, but it'll display something like, `Yay, more love for <PRIVATE CHANNEL>! Thanks, @whoever!`. Probably wouldn't hurt to put a check in for that.

File: src/scripts/angry-tock.test.js
Comment (Sgtpluck) [reply src/scripts/angry-tock.test.js:157]: actually this makes it a better long-term test maybe-- if shouty bot ever becomes location-aware then it should fail!

File: src/scripts/angry-tock.test.js
Comment (Sgtpluck) [reply src/scripts/angry-tock.test.js:181]: oh duh sorry my brain was melty at this point hah

File: src/scripts/angry-tock.test.js
Comment (Sgtpluck) [reply src/scripts/angry-tock.test.js:314]: i think there are definitely some test refactor opportunities in the future!

File: src/scripts/opm_status.js
Comment (mgwalker) [reply src/scripts/opm_status.js:42]: I'm legit not sure what errors this will catch. This is another old script and I just bonked it until it fit into the new API without thinking too much about it. 😂

File: src/scripts/love.js
Comment (Sgtpluck) [reply src/scripts/love.js:15]: ohhh ok didn't realize it was an auto-format!

File: src/scripts/travel-team.js
Comment (mgwalker) [reply src/scripts/travel-team.js:36]: If I remember correctly, the idea here is that it won't notify the same person within 3 hours. If you go to #travel on a Saturday and send several messages all in a row, it'll only respond to the first one. Then if you come back a few hours later, it'll respond again. Some concern about cluttering the channel with bot messages. This script needs more comments, I think. 😛

File: src/utils/slack.js
Comment (mgwalker) [reply src/utils/slack.js:23]: This should definitely be cached. It will change every time a new channel is created, but given how the bot uses it, it won't ever _meaningfully_ changed.

File: src/utils/slack.test.js
Comment (mgwalker) [reply src/utils/slack.test.js:43]: I'm not sure I understand the question. Do you mean getting the object that was passed to `msg.client.reactions.add`? We _could_ get that with `msg.client.reactions.add.mock.calls[0][0]`.

File: src/utils/tock.js
Comment (mgwalker) [reply src/utils/tock.js:35]: I KNOW RIGHT!

File: src/brain.test.js
Comment (mgwalker) [reply src/brain.test.js:20]: I'm not sure I understand what you mean here. Do you mean have the assertions in the `describe` instead of inside the `it`? Maybe we should sync up next week!

File: src/scripts/angry-tock.test.js
Comment (mgwalker) [reply src/scripts/angry-tock.test.js:118]: I like the way you're thinking. That's a better organization. I'm going to document these for #261 for post-migration refactor and cleanup, unless you think this should be done sooner rather than later!

File: src/scripts/federal-holidays-reminder.js
Comment (mgwalker) [reply src/scripts/federal-holidays-reminder.js:14]: Just gonna yoink `SUPPRESS_HERE` altogether. We can add it back someday if there's actually a call for it.

File: src/utils/slack.js
Comment (mgwalker) [reply src/utils/slack.js:23]: It turns out this doesn't need to be cached. It will only fetch the list of channels from Slack if the one it's looking for isn't already there, which should only happen once after load, because this function caches off a mapping of channel names to IDs itself.

File: src/brain.test.js
Comment (Sgtpluck) [reply src/brain.test.js:20]: yes -- happy to pair on this to make it clearer!

File: src/utils/slack.test.js
Comment (Sgtpluck) [reply src/utils/slack.test.js:43]: in general i prefer testing the state of objects rather than if a mocked function was called but i think i just misunderstood what was happening in this code trail! Thanks.

File: src/scripts/angry-tock.test.js
Comment (Sgtpluck) [reply src/scripts/angry-tock.test.js:124]: I can do a test cleanup pass after this gets merged in!

File: src/scripts/travel-team.js
Comment (Sgtpluck) [reply src/scripts/travel-team.js:33]: I think that makes sense, i think i was just wondering why 3 hours rather than 12!

File: src/brain.js
Comment (pburkholder) [reply src/brain.js:4]: Now that Redis will be AWS Elasticache, whether Postgres will be cheaper to maintain than Redis is a developer call, as the first 10 nodes of Redis are not charged in the cloud.gov pricing.

**Pull request merged**
